### ISSUE TITLE
CHERI capability pointer types should have an explicit byte size in DWARF debug info

### ISSUE THREAD

**Issue by bsdjhb** [CTSRD-CHERI/llvm#208]
GDB assumes that DW_TAG_pointer_type types have the same size as the address size in the containing compile unit unless an explicit size is specified via a DW_AT_byte_size attribute.  The current address size for CHERI binaries is still 8 bytes (64-bit addresses), so CHERI pointer types should be annotated with an explicit DW_AT_byte_size to indicate the pointer size in bytes.  For now I use a hack in GDB to override the size of pointer types for CHERI ABI binaries.   However, this hack can't work properly for hybrid code.  Properly annotating the size of pointer types will permit both hybrid and purecap code to be debugged without hacks in the debugger.

For example this C source:

```
extern int *myint;

int *myint;
```

Results in this DWARF debug info (readelf -w):

```
Contents of the .debug_info section:

  Compilation Unit @ offset 0x0:
   Length:        0x3d (32-bit)
   Version:       2
   Abbrev Offset: 0x0
   Pointer Size:  8
 <0><b>: Abbrev Number: 1 (DW_TAG_compile_unit)
    <c>   DW_AT_producer    : (indirect string, offset: 0x0): clang version 5.0.0 (git@github.com:CTSRD-CHERI/clang.git 520ebe16e7f8f96ccf30266dbab8d7d27c27f64d) (git@github.com:CTSRD-CHERI/llvm.git db1cf1fff6307cb2da2aa990cb1bc85bf0a171f3)
    <10>   DW_AT_language    : 12       (ANSI C99)
    <12>   DW_AT_name        : (indirect string, offset: 0xb3): test_dwarf_1.c
    <16>   DW_AT_stmt_list   : 0x0
    <1a>   DW_AT_comp_dir    : (indirect string, offset: 0xc2): /usr/home/john/work/johnsvn/test/hello
 <1><1e>: Abbrev Number: 2 (DW_TAG_variable)
    <1f>   DW_AT_name        : (indirect string, offset: 0xe9): myint
    <23>   DW_AT_type        : <0x34>
    <27>   DW_AT_external    : 1
    <28>   DW_AT_decl_file   : 1
    <29>   DW_AT_decl_line   : 3
    <2a>   DW_AT_location    : 9 byte block: 3 0 0 0 0 0 0 0 20         (DW_OP_addr: 20)
 <1><34>: Abbrev Number: 3 (DW_TAG_pointer_type)
    <35>   DW_AT_type        : <0x39>
 <1><39>: Abbrev Number: 4 (DW_TAG_base_type)
    <3a>   DW_AT_name        : (indirect string, offset: 0xef): int
    <3e>   DW_AT_encoding    : 5        (signed)
    <3f>   DW_AT_byte_size   : 4
 <1><40>: Abbrev Number: 0
```

The fix would be for the "int *" type at <34> to include a DW_AT_byte_size attribute indicating the size.  Alternatively we could perhaps treat the capability as a a new qualifier, but that would involve adding a new DWARF tag.  For now tagging the pointer type with the correct size seems the "simplest" approach.  DW_TAG_reference_type probably needs the same treatment for C++ references implemented via capabilities.

**Comment by kgudka** [comment]
Fixed in 51314d9a1d8f20917fe36a95a55ece7a31c48405 and CTSRD-CHERI/clang@51314d

**Comment by kgudka** [comment]
Fixed in commits 70b462defcb7cce1a3994008b2459916d92e1c4c and CTSRD-CHERI/clang@51314d9a1d8f20917fe36a95a55ece7a31c48405.
